<launch>
  <!-- size of map, change the size in x, y, z according to your application -->
  <arg name="map_size_x" value="50.0" />
  <arg name="map_size_y" value="50.0" />
  <arg name="map_size_z" value=" 5.1" />

  <arg name="init_x" default="-13.0" />
  <arg name="init_y" default="0.0" />
  <arg name="init_z" default="1.0" />

  <arg name="goal_x" default="10.0" />
  <arg name="goal_y" default="-10.0" />
  <arg name="goal_z" default="2.0" />

  <arg name="feature_filename_" default="feature_map_sample1.ply" />


  <!-- topic of your odometry such as VIO or LIO -->
  <arg name="odom_topic" value="/state_ukf/odom" />

  <!-- main algorithm params -->
  <include file="$(find exploration_manager)/launch/algorithm.xml">
    <arg name="feature_filename" value="$(find map_publisher)/data/$(arg feature_filename_)" />
    <arg name="map_size_x_" value="$(arg map_size_x)" />
    <arg name="map_size_y_" value="$(arg map_size_y)" />
    <arg name="map_size_z_" value="$(arg map_size_z)" />

    <!-- <arg name="box_min_x" value="-10.0"/>
    <arg name="box_min_y" value="-15.0"/>
    <arg name="box_min_z" value=" 0.0"/>
    <arg name="box_max_x" value="10.0"/>
    <arg name="box_max_y" value="15.0"/>
    <arg name="box_max_z" value=" 2.0"/> -->

    <arg name="box_min_x" value="-15.0" />
    <arg name="box_min_y" value="-15.0" />
    <arg name="box_min_z" value="0.5" />
    <arg name="box_max_x" value="15.0" />
    <arg name="box_max_y" value="15.0" />
    <arg name="box_max_z" value=" 4.0" />

    <arg name="odometry_topic" value="$(arg odom_topic)" />
    <!-- sensor pose: transform of camera frame in the world frame -->
    <arg name="sensor_pose_topic" value="/pcl_render_node/sensor_pose" />
    <!-- depth topic: depth image, 640x480 by default -->
    <!-- cloud topic: point cloud measurement  -->
    <!-- subscribe ONLY TO ONE of the two topics -->
    <arg name="depth_topic" value="/pcl_render_node/depth" />
    <arg name="cloud_topic" value="/pcl_render_node/cloud" />

    <!-- intrinsic params of the depth camera -->
    <arg name="cx" value="321.04638671875" />
    <arg name="cy" value="243.44969177246094" />
    <arg name="fx" value="387.229248046875" />
    <arg name="fy" value="387.229248046875" />

    <!-- maximum velocity and acceleration the drone will reach -->
    <arg name="max_vel" value="0.4" />
    <arg name="max_acc" value="0.15" />
    <!-- <arg name="max_vel" value="1.0" />
    <arg name="max_acc" value="0.4" /> -->

    <!-- ADD: maximum yaw rate the drone will reach -->
    <arg name="max_yaw_rate" value="0.6" />

    <arg name="min_feature_num_act" value="10" />
    <arg name="min_covisible_feature_num_act" value="6" />

    <arg name="min_feature_num_plan" value="15" />
    <arg name="min_covisible_feature_num_plan" value="10" />

    <!-- 1: use 2D Nav Goal to select goal  -->
    <!-- 2: use global waypoints below  -->
    <arg name="flight_type" value="2" />

    <arg name="point_num" value="4" />

    <arg name="point0_x" value="$(arg goal_x)" />
    <arg name="point0_y" value="$(arg goal_y)" />
    <arg name="point0_z" value="$(arg goal_z)" />

    <arg name="point1_x" value="-8.0" />
    <arg name="point1_y" value="-1.5" />
    <arg name="point1_z" value="1.0" />

    <arg name="point2_x" value="-6.0" />
    <arg name="point2_y" value="-1.7" />
    <arg name="point2_z" value="1.0" />

    <arg name="point3_x" value="-3.0" />
    <arg name="point3_y" value="0.0" />
    <arg name="point3_z" value="1.0" />
  </include>

  <!-- trajectory server -->
  <node pkg="plan_manage" name="traj_server" type="traj_server" output="screen">
    <remap from="/position_cmd" to="planning/pos_cmd" />

    <remap from="/odom_world" to="$(arg odom_topic)" />
    <remap from="/odom_camera" to="/pcl_render_node/sensor_pose" />
    <param name="traj_server/time_forward" value="1.5" type="double" />
    <param name="traj_server/pub_traj_id" value="4" type="int" />
    <param name="traj_server/init_x" value="$(arg init_x)" type="double" />
    <param name="traj_server/init_y" value="$(arg init_y)" type="double" />
    <param name="traj_server/init_z" value="$(arg init_z)" type="double" />

    <param name="perception_utils/top_angle" value="0.56125" type="double" />
    <param name="perception_utils/left_angle" value="0.69222" type="double" />
    <param name="perception_utils/right_angle" value="0.68901" type="double" />
    <param name="perception_utils/max_dist" value="4.5" type="double" />
    <param name="perception_utils/vis_dist" value="1.0" type="double" />

  </node>

  <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen">
    <remap from="~odom" to="$(arg odom_topic)" />
    <remap from="~goal" to="/move_base_simple/goal" />
    <remap from="~traj_start_trigger" to="/traj_start_trigger" />
    <param name="waypoint_type" value="manual-lonely-waypoint" />
  </node>

  <!-- use simulator -->
  <include file="$(find exploration_manager)/launch/simulator.xml">
    <arg name="map_size_x_" value="$(arg map_size_x)" />
    <arg name="map_size_y_" value="$(arg map_size_y)" />
    <arg name="map_size_z_" value="$(arg map_size_z)" />

    <arg name="init_x" value="$(arg init_x)" />
    <arg name="init_y" value="$(arg init_y)" />
    <arg name="init_z" value="$(arg init_z)" />

    <arg name="odometry_topic" value="$(arg odom_topic)" />
    <arg name="c_num" value="0" />
    <arg name="p_num" value="130" />
  </include>

</launch>